packages/attestation-service/src/attestation.ts

```diff
@@ -1,38 +1,128 @@
+
+  // TODO: Check with the new Accounts.sol
+  if (address.toLowerCase() !== request.issuer.toLowerCase()) {
+    throw new Error(`Mismatching issuer, I am ${address}`)
+  }
+
+  const attestations = await kit.contracts.getAttestations()
+  const state = await attestations.getAttestationState(
+    request.phoneNumber,
+    request.account,
+    request.issuer
+  )
+
+  console.info(state)
+  if (state.attestationState !== AttestationState.Incomplete) {
+    throw new Error('No incomplete attestation found')
+  }
+
+  // TODO: Check expiration
+  return
+}
+
+async function validateAttestation(
+  attestationRequest: AttestationRequest,
+  attestationCode: string
+) {
+  const attestations = await kit.contracts.getAttestations()
+  const isValid = await attestations.validateAttestationCode(
+    attestationRequest.phoneNumber,
+    attestationRequest.account,
+    attestationRequest.issuer,
+    attestationCode
+  )
+  if (!isValid) {
+    throw new Error('Valid attestation could not be provided')
+  }
+  return
+}
+
+function signAttestation(phoneNumber: E164Number, account: Address) {
+  const signature = attestToIdentifier(phoneNumber, account, getAttestationKey())
 
   return SignatureUtils.serializeSignature(signature)
 }
 
 function toBase64(str: string) {
-  return Buffer.from(str, 'hex').toString('base64')
+  return Buffer.from(str.slice(2), 'hex').toString('base64')
 }
 
 function createAttestationTextMessage(attestationCode: string) {
   return `<#> ${toBase64(attestationCode)} ${process.env.APP_SIGNATURE}`
 }
 
-export async function handleAttestationRequest(req: express.Request, res: express.Response) {
-  // TODO: Should parse request appropriately
-
-  // TODO: Should validate request here
-  // const attestations = await kit.contracts.getAttestations()
-
-  // Produce attestation
-  const attestationCode = signAttestation(req.body.phoneNumber, req.body.account)
-  const textMessage = createAttestationTextMessage(attestationCode)
+export async function handleAttestationRequest(
+  _req: express.Request,
+  res: express.Response,
+  attestationRequest: AttestationRequest
+) {
+  let attestationCode
+  try {
+    await validateAttestationRequest(attestationRequest)
+    attestationCode = signAttestation(attestationRequest.phoneNumber, attestationRequest.account)
+    await validateAttestation(attestationRequest, attestationCode)
+  } catch (error) {
+    console.error(error)
+    res.status(422).json({ success: false, error: error.toString() })
+    return
+  }
 
-  // Send the SMS
-  await retryAsyncWithBackOff(sendSms, 10, [req.body.phoneNumber, textMessage], 1000)
+  try {
+    const textMessage = createAttestationTextMessage(attestationCode)
+    await persistAttestationRequest(
+      attestationRequest.phoneNumber,
+      attestationRequest.account,
+      attestationRequest.issuer
+    )
+    await retryAsyncWithBackOff(sendSms, 10, [attestationRequest.phoneNumber, textMessage], 1000)
```

Comment:
Nit: wrapping the phone number and txt message in an array seems odd to me

Reply:
That's our current interface into `retryAsyncWithBackOff` which I don't think we have much alternative to unless we convert all our functions to only take one argument?

Reply:
if we make change to signature to move up the delay param and then have `...args` for the rest, we won't need to wrap the params. But it's nbd